### PR TITLE
Hats event listeners

### LINKED ISSUE #2112: Update Roles page data when roles edits are executed
In the data loaders, set up some listeners for the different Hats changes.

### PR DESCRIPTION
**Pull request by adamgall**
closes #2112 
See [this review on an old PR](https://github.com/decentdao/decent-interface/pull/2111#pullrequestreview-2167861285) (which was the birthplace of this code) as a preamble for this PR.

### CODE REVIEW

File: src/hooks/DAO/useHatsListeners.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    if (!publicClient || !daoAddress) return;
+
+    const hatsContract = getContract({
+      abi: HatsAbi,
+      address: getAddress(hatsProtocol),
+      client: publicClient,
+    });
+
+    const unwatchHatCreated = hatsContract.watchEvent.HatCreated({
+      onLogs: log => handleHatCreated(log[0].args),
```
Comment (adamgall) [inline src/hooks/DAO/useHatsListeners.ts:100]: `log` is an array (should be called `logs`). Make sure you "handle" each log event in the array.

File: src/hooks/DAO/useHatsListeners.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+  const isNotFromThisSafesTree = useCallback(
+    (hatId: bigint | undefined) => {
+      if (!hatId) return true;
+
+      const treeId = hatIdToTreeId(hatId);
+      return treeId !== hatsTreeId;
+    },
+    [hatsTreeId],
+  );
```
Comment (adamgall) [inline src/hooks/DAO/useHatsListeners.ts:33]: Too bad we can't do event-layer filtering huh... I don't love this, we'll be getting every event from every Hat in our client here.

Comment (Da-Colon) [reply src/hooks/DAO/useHatsListeners.ts:33]: o dang.. hmm that is unfortunate dang.

File: src/hooks/DAO/useHatsListeners.ts (hunk 1 above)
Comment (Da-Colon) [reply src/hooks/DAO/useHatsListeners.ts:100]: agreed

File: src/hooks/DAO/useHatsListeners.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+  // @todo Each of these `handle*` functions can probably be updated to more granularly update a specific node on the hats tree. If it turns out to be impossible or unnecessary, them these functions are entirely bloat and `getHatsTree` should be called directly in each listener's `onLogs` callback.
```
Comment (Da-Colon) [inline src/hooks/DAO/useHatsListeners.ts:35]: agreed with the approach

**Review by Da-Colon — APPROVED** [review]

**Pull request closed without merging**